jujusvg.go

```diff
@@ -10,11 +12,51 @@ import (
 	"gopkg.in/juju/charm.v5-unstable"
 )
 
-// NewFromBundle returns a new Canvas that can be used
+func NewFromBundleEmbedIcons(b *charm.BundleData, iconURL func(*charm.Reference) string) (*Canvas, error) {
+	alreadyFetched := make(map[string]bool)
+	icons := make(map[string]string)
+	for _, serviceData := range b.Services {
+		charmId, err := charm.ParseReference(serviceData.Charm)
+		if err != nil {
+			return nil, errgo.Notef(err, "cannot parse charm %q", serviceData.Charm)
+		}
+		if _, ok := alreadyFetched[charmId.Path()]; ok == false {
+			alreadyFetched[charmId.Path()] = true
+			icon, err := fetchIcon(iconURL(charmId))
```

Comment:
@rogpeppe Done, but this is my first attempt at writing concurrent go code, so I welcome feedback!